docs/design/generics/details.md

```diff
@@ -3785,24 +3806,58 @@ external impl forall [T:! Printable] Vector(T) as Printable {
 }
 ```
 
-To include these `impl` definitions inline in a `class` definition, include a
-`forall` clause with a more-specific type between the `impl` and `as` keywords.
+Note that no `forall` clause or type may be specified when declaring an `impl`
+with the `extend` keyword:
 
 ```
 class Array(T:! type, template N:! i64) {
-  impl forall [P:! Printable] Array(P, N) as Printable { ... }
+  // ❌ Illegal: nothing allowed before `as` after `extend impl`:
+  extend impl forall [P:! Printable] Array(P, N) as Printable { ... }
 }
 ```
 
-All internal `impl` declarations in the body of a `class` definition must be for
-the class being defined. It is an error to declare `impl i32 as Printable`
-inside `class Array`.
+**Note:** This was changed in
+[proposal #2760](https://github.com/carbon-language/carbon-lang/pull/2760).
+
+Instead, the class can declare aliases to members of the interface. Those
+aliases will only be usable when the type implements the interface.
+
+```
+class Array(T:! type, template N:! i64) {
+  alias Print = Printable.Print;
+}
+impl forall [P:! Printable] Array(P, N) as Printable { ... }
+
+impl String as Printable { ... }
+var can_print: Array(String, 2) = ("Hello ", "world");
+// ✅ Allowed: `can_print.Print` resolves to
+// `can_print.(Printable.Print)`, which exists as long as
+// `Array(String, 2) impls Printable`, which exists since
+// `String impls Printable`.
+can_print.Print();
+
+var no_print: Array(Unprintable, 2) = ...;
+// ❌ Illegal: `no_print.Print` resolves to
+// `no_print.(Printable.Print)`, but there is no
+// implementation of `Printable` for `Array(Unprintable, 2)`
+// as long as `Unprintable` doesn't implement `Printable`.
+no_print.Print();
+```
 
-It is legal to add the keyword `external` before the `impl` keyword to switch to
-an external impl defined lexically within the class scope. Inside the scope,
-both `P` and `T` refer to the same type, but `P` has the type-of-type of
-`Printable` and so has a `Print` member. The relationship between `T` and `P` is
-as if there was a `where P == T` clause.
+It is still legal to declare or define an external impl lexically with the class
+scope, as in:
+
+```
+class Array(T:! type, template N:! i64) {
+  // ✅ Allowed: external impl defined in class scope may use `forall`
+  // and/or specify a type.
```

Comment:
```suggestion
  // or specify a type.
```

Style encourages rewriting to avoid `and/or`. In this case, `impl X forall Y as Z` is invalid, so maybe we can drop `and`? ("may specify a type or use `forall`" might read more easily, too)
https://developers.google.com/style/word-list#and-or